File: src/test/java/kitchenpos/application/MenuGroupServiceMockTest.java
package kitchenpos.application;

import kitchenpos.ApplicationMockTest;
import kitchenpos.domain.MenuGroup;
import kitchenpos.domain.MenuGroupRepository;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.NullAndEmptySource;
import org.mockito.InjectMocks;
import org.mockito.Mock;

import java.util.List;

import static kitchenpos.fixture.MenuGroupFixture.menuGroupCreateRequest;
import static kitchenpos.fixture.MenuGroupFixture.menuGroupResponse;
import static kitchenpos.fixture.MenuGroupFixture.이름_추천메뉴;
import static kitchenpos.fixture.MenuGroupFixture.이름_한마리메뉴;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.junit.jupiter.api.Assertions.assertAll;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.when;

@DisplayName("메뉴그룹 서비스 테스트")
@ApplicationMockTest
class MenuGroupServiceMockTest {
    @Mock
    private MenuGroupRepository menuGroupRepository;

    @InjectMocks
    private MenuGroupService menuGroupService;

    @DisplayName("메뉴그룹 등록한다.")
    @Nested
    class MenuGroupCreate {
        @DisplayName("[성공] 등록")
        @Test
        void success() {
            // given
            MenuGroup request = menuGroupCreateRequest(이름_추천메뉴);
            when(menuGroupRepository.save(any(MenuGroup.class))).thenReturn(menuGroupResponse(이름_추천메뉴));

            // when
            MenuGroup result = menuGroupService.create(request);

            // then
            assertAll(
                    () -> assertThat(result.getId()).isNotNull(),
                    () -> assertThat(result.getName()).isEqualTo(이름_추천메뉴)
            );
        }

        @DisplayName("[실패] 이름은 공백일 수 없다.")
        @NullAndEmptySource
        @ParameterizedTest
        void fail1(String name) {
            // given
            MenuGroup request = menuGroupCreateRequest(name);

            // when
            // then
            assertThatThrownBy(() -> menuGroupService.create(request))
                    .isInstanceOf(IllegalArgumentException.class);
        }
    }

    @DisplayName("메뉴그룹 목록을 볼 수 있다")
    @Nested
    class MenuGroupList {
        @DisplayName("[성공] 목록보기")
        @Test
        void getMenuGroups() {
            // given
            List<MenuGroup> 메뉴그룹_목록 = List.of(menuGroupResponse(이름_추천메뉴), menuGroupResponse(이름_한마리메뉴));
            when(menuGroupRepository.findAll()).thenReturn(메뉴그룹_목록);

            // when
            List<MenuGroup> result = menuGroupService.findAll();

            // then
            assertThat(result).extracting("name").containsExactly(이름_추천메뉴, 이름_한마리메뉴);
        }
    }
}

File: src/test/java/kitchenpos/application/MenuGroupServiceMockTest.kt
package kitchenpos.application

import io.kotest.core.spec.style.BehaviorSpec
import io.kotest.matchers.shouldBe
import io.kotest.matchers.shouldNotBe
import kitchenpos.domain.FakeMenuGroupRepository
import kitchenpos.domain.MenuGroup
import org.junit.jupiter.api.assertThrows

private val menuGroupRepository = FakeMenuGroupRepository()
private val menuGroupService = MenuGroupService(menuGroupRepository)

class MenuGroupServiceMockTest : BehaviorSpec({
    given("메뉴 그룹을 생성할 때") {
        `when`("입력 값이 정상이면") {
            val newMenuGroup = MenuGroup().apply { name = "메뉴 그룹" }

            then("정상적으로 생성된다.") {
                with(menuGroupService.create(newMenuGroup)) {
                    this.id shouldNotBe null
                    this.name shouldBe "메뉴 그룹"
                }
            }
        }

        `when`("이름이 null 이면") {
            val newMenuGroup = MenuGroup().apply { name = null }

            then("예외가 발생한다.") {
                assertThrows<IllegalArgumentException> {
                    menuGroupService.create(newMenuGroup)
                }
            }
        }

        `when`("이름이 빈 문자열이면") {
            val newMenuGroup = MenuGroup().apply { name = "" }

            then("예외가 발생한다.") {
                assertThrows<IllegalArgumentException> {
                    menuGroupService.create(newMenuGroup)
                }
            }
        }
    }

    given("메뉴 그룹을 조회할 떄") {
        `when`("메뉴 그룹이 존재하지 않으면") {
            menuGroupRepository.clear()

            then("빈 목록을 반환한다.") {
                with(menuGroupService.findAll()) {
                    this.size shouldBe 0
                }
            }
        }
    }
})
